File: appcore/src/main/java/mobi/cangol/mobile/service/route/RouteServiceImpl.java
package mobi.cangol.mobile.service.route;

import android.app.Activity;
import android.app.Application;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import androidx.fragment.app.Fragment;

import java.util.HashMap;
import java.util.Map;

import mobi.cangol.mobile.logging.Log;
import mobi.cangol.mobile.service.Service;
import mobi.cangol.mobile.service.ServiceProperty;

/**
 * Created by xuewu.wei on 2018/10/15.
 */
@Service("RouteService")
class RouteServiceImpl implements RouteService {
    private static final String TAG = "RouteService";
    private ServiceProperty mServiceProperty = null;
    private Map<String, Class<?>> mRouteMap = null;
    private OnNavigation mOnNavigation;

    private boolean mDebug = false;

    @Override
    public void onCreate(Application context) {
        mRouteMap = new HashMap<>();
    }

    @Override
    public void setDebug(boolean mDebug) {
        this.mDebug = mDebug;
    }

    @Override
    public void init(ServiceProperty serviceProperty) {
        this.mServiceProperty = serviceProperty;
    }

    @Override
    public ServiceProperty getServiceProperty() {
        return mServiceProperty;
    }

    @Override
    public ServiceProperty defaultServiceProperty() {
        return new ServiceProperty(TAG);
    }

    @Override
    public String getName() {
        return TAG;
    }

    @Override
    public void onDestroy() {
        mRouteMap.clear();
    }


    @Override
    public void registerByAnnotation(Class<?> clazz) {
        if (clazz.isAnnotationPresent(Route.class)) {
            final Route route = clazz.getAnnotation(Route.class);
            register(route.path(), clazz);
        } else {
            throw new IllegalStateException(clazz + " is not Annotation Route");
        }
    }

    @Override
    public void unregisterByAnnotation(Class<?> clazz) {
        if (clazz.isAnnotationPresent(Route.class)) {
            final Route route = clazz.getAnnotation(Route.class);
            unregister(route.path());
        } else {
            throw new IllegalStateException(clazz + " is not Annotation Route");
        }
    }

    @Override
    public void register(String path, Class<?> clazz) {
        if (!mRouteMap.containsKey(path)) {
            if(mDebug)Log.i(TAG, "registerRoute " + path + "--->" + clazz.getName());
            mRouteMap.put(path, clazz);
        } else {
            Log.i(TAG, path + " is registered");
        }
    }

    @Override
    public void unregister(String path) {
        if (mRouteMap.containsKey(path)) {
            if(mDebug) Log.i(TAG, "unregisterRoute " + path);
            mRouteMap.remove(path);
        } else {
            Log.i(TAG, path + " is not registered");
        }
    }

    @Override
    public RouteBuilder build(String path) {
        return new RouteBuilder(this, path);
    }

    @Override
    public void handleIntent(Context context,Intent intent) {
        Uri uri = intent.getData();
        if(uri!=null){
            Log.i(TAG, "handleIntent uri="+uri+",host="+uri.getHost()+",path="+uri.getPath());
            Bundle bundle=new Bundle();
            for (String key : uri.getQueryParameterNames()) {
                bundle.putString(key,uri.getQueryParameter(key));
            }
            this.handleNavigation(uri.getHost()+uri.getPath(),
                    bundle,
                    context,
                    uri.getBooleanQueryParameter("newStack",false),0,-1);
        } else {
            Log.i(TAG, "intent.getData()==null isn't handle");
        }
    }

    @Override
    public void registerNavigation(OnNavigation onNavigation) {
        this.mOnNavigation = onNavigation;
    }

    public void handleNavigation(String path, Bundle bundle, Context context, boolean newStack, int flags, int requestCode) {
        final Class<?> clazz = mRouteMap.get(path);
        if (clazz != null) {
            if (Activity.class.isAssignableFrom(clazz)) {
                final Intent intent = new Intent(context, clazz);
                intent.putExtras(bundle);
                if(flags>0)intent.setFlags(flags);
                if(context instanceof Activity){
                    ((Activity)context).startActivityForResult(intent,requestCode);
                }else{
                    context.startActivity(intent);
                }
            } else if (Fragment.class.isAssignableFrom(clazz)) {
                this.mOnNavigation.toFragment((Class<? extends Fragment>) clazz, bundle,newStack);
            } else {
                Log.e(TAG, path+" not navigation");
                this.mOnNavigation.notFound(path);
            }
        }else{
            Log.e(TAG, path+" not register");
            this.mOnNavigation.notFound(path);
        }
    }
}

File: appcore/src/main/java/mobi/cangol/mobile/service/route/RouteServiceImpl.kt
package mobi.cangol.mobile.service.route

import android.app.Activity
import android.app.Application
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.support.v4.app.Fragment
import mobi.cangol.mobile.logging.Log
import mobi.cangol.mobile.service.Service
import mobi.cangol.mobile.service.ServiceProperty
import java.util.*

/**
 * Created by xuewu.wei on 2018/10/15.
 */
@Service("RouteService")
class RouteServiceImpl : RouteService {
    private var mServiceProperty = ServiceProperty(TAG)
    private var mRouteMap: MutableMap<String, Class<*>>? = null
    private var mOnNavigation: OnNavigation? = null
    private var mDebug = false

    override fun onCreate(context: Application) {
        mRouteMap = HashMap()
    }

    override fun setDebug(mDebug: Boolean) {
        this.mDebug = mDebug
    }

    override fun init(serviceProperty: ServiceProperty) {
        this.mServiceProperty = serviceProperty
    }

    override fun getServiceProperty(): ServiceProperty {
        return mServiceProperty
    }

    override fun defaultServiceProperty(): ServiceProperty {
        return mServiceProperty
    }

    override fun getName(): String {
        return TAG
    }

    override fun onDestroy() {
        mRouteMap!!.clear()
    }


    override fun register(clazz: Class<*>) {
        if (clazz.isAnnotationPresent(Route::class.java)) {
            register(clazz.getAnnotation(Route::class.java)!!.path, clazz)
        } else {
            throw IllegalStateException("$clazz is not Annotation Route")
        }
    }

    override fun register(path: String, clazz: Class<*>) {
        if (!mRouteMap!!.containsKey(path)) {
            if (mDebug) Log.i(TAG, "registerRoute " + path + "--->" + clazz.name)
            mRouteMap!![path] = clazz
        } else {
            Log.i(TAG, "$path is registered")
        }
    }

    override fun unregister(path: String) {
        if (mRouteMap!!.containsKey(path)) {
            if (mDebug) Log.i(TAG, "unregisterRoute $path")
            mRouteMap!!.remove(path)
        } else {
            Log.i(TAG, "$path is not registered")
        }
    }

    override fun build(path: String): RouteBuilder {
        return RouteBuilder(this, path)
    }

    override fun registerNavigation(onNavigation: OnNavigation) {
        this.mOnNavigation = onNavigation
    }

    fun getClassByPath(path: String): Class<*>? {
        return mRouteMap!![path]
    }

    fun handleNavigation(clazz: Class<*>, bundle: Bundle, context: Context) {
        if (clazz.superclass == Activity::class.java) {
            this.mOnNavigation!!.toActivity(navigationActivity(clazz, bundle, context))
        } else if (clazz.superclass == Fragment::class.java) {
            this.mOnNavigation!!.toFragment(navigationFragment(clazz, bundle, context))
        } else {
            Log.i(TAG, " not navigation")
        }
    }

    fun navigationActivity(clazz: Class<*>, bundle: Bundle, context: Context): Intent {
        val intent = Intent(context, clazz)
        intent.putExtras(bundle)
        return intent
    }

    fun navigationFragment(clazz: Class<*>, bundle: Bundle, context: Context): Fragment {
        return Fragment.instantiate(context, clazz.name, bundle)
    }

    companion object {
        private const val TAG = "RouteService"
    }

}
